clients/githubrepo/branches.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2021 Security Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package githubrepo
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/google/go-github/v38/github"
+	"github.com/shurcooL/githubv4"
+
+	"github.com/ossf/scorecard/v2/clients"
+	sce "github.com/ossf/scorecard/v2/errors"
+)
+
+const (
+	refsToAnalyze = 30
+	refPrefix     = "refs/heads/"
+)
+
+type refUpdateRule struct {
+	AllowsDeletions              *githubv4.Boolean
+	AllowsForcePushes            *githubv4.Boolean
+	RequiredApprovingReviewCount *githubv4.Int
+	RequiresCodeOwnerReviews     *githubv4.Boolean
+	RequiresLinearHistory        *githubv4.Boolean
+	RequiredStatusCheckContexts  []githubv4.String
+}
+
+type branchProtectionRule struct {
+	DismissesStaleReviews      *githubv4.Boolean
+	IsAdminEnforced            *githubv4.Boolean
+	RequiresStrictStatusChecks *githubv4.Boolean
+}
+
+type branch struct {
+	Name                 *githubv4.String
+	RefUpdateRule        *refUpdateRule
+	BranchProtectionRule *branchProtectionRule
+}
+
+type branchesData struct {
+	Repository struct {
+		DefaultBranchRef branch
+		Refs             struct {
+			Nodes []branch
+		} `graphql:"refs(first: $refsToAnalyze, refPrefix: $refPrefix)"`
+	} `graphql:"repository(owner: $owner, name: $name)"`
+}
+
+type branchesHandler struct {
+	ghClient         *github.Client
+	graphClient      *githubv4.Client
+	data             *branchesData
+	once             *sync.Once
+	ctx              context.Context
+	errSetup         error
+	owner            string
+	repo             string
+	defaultBranchRef *clients.BranchRef
+	branches         []*clients.BranchRef
+}
+
+func (handler *branchesHandler) init(ctx context.Context, owner, repo string) {
+	handler.ctx = ctx
+	handler.owner = owner
+	handler.repo = repo
+	handler.errSetup = nil
+	handler.once = new(sync.Once)
+}
+
+func (handler *branchesHandler) setup() error {
+	handler.once.Do(func() {
+		vars := map[string]interface{}{
+			"owner":         githubv4.String(handler.owner),
+			"name":          githubv4.String(handler.repo),
+			"refsToAnalyze": githubv4.Int(refsToAnalyze),
+			"refPrefix":     githubv4.String(refPrefix),
+		}
+		handler.data = new(branchesData)
+		if err := handler.graphClient.Query(handler.ctx, handler.data, vars); err != nil {
+			handler.errSetup = sce.Create(sce.ErrScorecardInternal, fmt.Sprintf("githubv4.Query: %v", err))
```

Comment:
since the repo implementation returns a `ErrScorecardInternal`, we can assume in the checks that errors from any `repo.*` functions can be returned as-is. Correct?

Reply:
As of now, yes. But I think its best we don't keep this assumption. The `sce.Err*` errors should be returned by `checks` package. Internal packages like `gtihubrepo` should be free to return other internal errors.

Reply:
if we add `ErrScorecardInternal` in checks and the repo does too, we end up with multiple `ErrScorecardInternal` wrapping one another. Is there a better way to do it?

Reply:
Let's file an issue for cleaning this up because I think as of today, there is no clear guideline. IMO, only `checks` and `pkg` packages of Scorecard should return these `Internal` errors.

Reply:
there's one already https://github.com/ossf/scorecard/issues/783